docs/develop/porting.rst

```diff
@@ -0,0 +1,325 @@
+    #include <alloca.h>
+
+    #define MICROPY_HW_BOARD_NAME "example-board""
+    #define MICROPY_HW_MCU_NAME   "unknown-cpu"
+
+    #ifdef __linux__
+    #define MICROPY_MIN_USE_STDOUT (1)
+    #endif
+
+    #ifdef __thumb__
+    #define MICROPY_MIN_USE_CORTEX_CPU (1)
+    #define MICROPY_MIN_USE_STM32_MCU (1)
+    #endif
+
+    #define MP_STATE_PORT MP_STATE_VM
+
+    #define MICROPY_PORT_ROOT_POINTERS \
+        const char *readline_hist[8];
+
+    #endif
+
+This configuration file contains machine-specific configurations including aspects like if different
+MicroPython features should be enabled i.e ``#define MICROPY_ENABLE_GC (1) ``. Making this Setting
+``0`` disables the feature.
+
+Other configurations include type definitions, root pointers, board name, microcontroller name
+etc.
+
+Similarly, an example ``mphalport.h`` file looks like this:
+
+.. code-block:: c
+
+    static inline mp_uint_t mp_hal_ticks_ms(void) { return 0; }
+    static inline void mp_hal_set_interrupt_char(char c) {}
+    void mp_hal_cmsis_raise(int32_t status);
+
+Support for standard input/output
+----------------------------------
+
+Now add a file ``uart_core.c`` to allow for input and output:
+
+.. code-block:: c
+
+    #include <unistd.h>
+    #include "py/mpconfig.h"
+
+
+    // Receive single character
+    int mp_hal_stdin_rx_chr(void) {
+        unsigned char c = 0;
+        #if MICROPY_MIN_USE_STDOUT
+        int r = read(STDIN_FILENO, &c, 1);
+        (void)r;
+        #endif
+        return c;
+    }
+
+    // Send string of given length
+    void mp_hal_stdout_tx_strn(const char *str, mp_uint_t len) {
+        #if MICROPY_MIN_USE_STDOUT
+        int r = write(STDOUT_FILENO, str, len);
+        (void)r;
+        #endif
+    }
+
+These input and output functions have to be modified depending on the
+specific board API. This example uses the standard input/output stream.
+
+Adding a module to a port
+--------------------------
+
+To support an existing MicroPython module like ``pyb``, first, add the module
+definition in a file ``modpyb.c``:
+
+.. code-block:: c
+
+    #include "py/mpconfig.h"
+    #include "py/obj.h"
+
+    STATIC const mp_map_elem_t pyb_module_globals_table[] = {
+        { MP_OBJ_NEW_QSTR(MP_QSTR___name__), MP_OBJ_NEW_QSTR(MP_QSTR_pyb) },
+    };
+
+    STATIC MP_DEFINE_CONST_DICT(pyb_module_globals, pyb_module_globals_table);
+
+    const mp_obj_module_t pyb_module = {
+        .base = { &mp_type_module },
+        .globals = (mp_obj_dict_t*)&pyb_module_globals,
+    };
+
+Modify the configuration file ``mpconfigport.h`` accordingly:
+
+.. code-block:: c
+
+    // extra built-in modules to add to the list of known ones
+    extern const struct _mp_obj_module_t pyb_module;
+
+    #define MICROPY_PORT_BUILTIN_MODULES \
+        { MP_OBJ_NEW_QSTR(MP_QSTR_pyb), (mp_obj_t)&pyb_module },
+
+    // extra constants
+    #define MICROPY_PORT_CONSTANTS \
+        { MP_OBJ_NEW_QSTR(MP_QSTR_pyb), (mp_obj_t)&pyb_module }, \
```

Comment:
This extra constants bit is only needed if the module wants to export constants for the parser to fold (optimise).  For this minimal port/minimal example I think it's simpler to just remove this bit.